reactioncommerce/reaction
# Add cancelOrderItem mutation

Resolves #2382
Impact: **minor**  
Type: **feature**

## Changes
A new `cancelOrderItem` mutation allows you to cancel a single order item or a partial quantity of a single order item.

The operator UI does not yet use this. That will be done in a separate PR, at which time the `cancelOrder` Meteor method will also be removed.

- `cancelOrderItem` can be called internally to synchronize from an external system by setting `context.isInternalCall` to `true`
- `cancelOrderItem` can be called through GraphQL by any user with "orders" permission for the shop that owns the order
- `cancelOrderItem` can be called through GraphQL by the user who placed the order. In this case, the cancellation will only succeed if both the item status and the order status are "new".
- Unlike with the existing Meteor method, neither restocking nor refunds are handled by this mutation. These are separate things with separate systems, so we'll now expect clients to separately ask for restocking or refunds as necessary.
- When the last item in a group is canceled, the group is also canceled. When the last group in an order is canceled, the order is also canceled.

## Breaking changes
None

## Testing
1. Place an order.
1. Verify that the mutation works as described.
1. Verify that you can cancel an order item when authenticated as a user with "orders" role or as the user who placed the order, but not when authenticated as someone else or when not authenticated.
